src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/specification/ReferralSpecifications.kt

```diff
@@ -0,0 +1,119 @@
+     * ReferralConcluder.kt class.
+     */
+    fun cancelled(): Specification<Referral> {
+      return Specification<Referral> { root, query, cb ->
+        cb.and(
+          cb.isNotNull(root.get<OffsetDateTime>("endRequestedAt")),
+          cb.isNotNull(root.get<OffsetDateTime>("concludedAt")),
+          root.join<Referral, EndOfServiceReport>("endOfServiceReport", JoinType.LEFT).isNull
+        )
+      }
+    }
+
+    fun withSPAccess(contracts: Set<DynamicFrameworkContract>): Specification<Referral> {
+      return Specification<Referral> { root, query, cb ->
+        val interventionJoin = root.join<Referral, Intervention>("intervention", JoinType.INNER)
+        val dynamicContractJoin = interventionJoin.join<Intervention, DynamicFrameworkContract>("dynamicFrameworkContract", JoinType.LEFT)
+        dynamicContractJoin.`in`(contracts)
+      }
+    }
+
+    fun matchingPrimeProviderReferrals(serviceProviders: Set<ServiceProvider>): Specification<Referral> {
+      return Specification<Referral> { root, query, cb ->
+        val interventionJoin = root.join<Referral, Intervention>("intervention", JoinType.INNER)
+        val dynamicContractJoin = interventionJoin.join<Intervention, DynamicFrameworkContract>("dynamicFrameworkContract", JoinType.LEFT)
+        query.distinct(true)
+        dynamicContractJoin.get<ServiceProvider>("primeProvider").`in`(serviceProviders)
+      }
+    }
+
+    fun matchingSubContractorReferrals(serviceProviders: Set<ServiceProvider>): Specification<Referral> {
+      return Specification<Referral> { root, query, cb ->
+        val interventionJoin = root.join<Referral, Intervention>("intervention", JoinType.INNER)
+        val dynamicContractJoin = interventionJoin.join<Intervention, DynamicFrameworkContract>("dynamicFrameworkContract", JoinType.INNER)
+        val serviceProviderJoin = dynamicContractJoin.joinSet<DynamicFrameworkContract, ServiceProvider>("subcontractorProviders", JoinType.LEFT)
+        query.distinct(true)
+        serviceProviderJoin.get<AuthGroupID>("id").`in`(serviceProviders.map { it.id })
+      }
+    }
+
+    fun createdBy(authUser: AuthUser): Specification<Referral> {
+      return Specification<Referral> { root, query, cb -> cb.equal(root.get<String>("createdBy"), authUser) }
+    }
+
+    fun matchingServiceUserReferrals(serviceUserCRNs: List<String>): Specification<Referral> {
+      return Specification<Referral> { root, query, cb -> root.get<String>("serviceUserCRN").`in`(serviceUserCRNs) }
+    }
+
+    /**
+     * This is performing the equivalent of:
+     *
+     * SELECT r FROM referral r
+     * inner join referral_assignments ra
+     *    inner join auth_user au on ra.assigned_to_id = au.id
+     *  on ra.referral_id = r.id
+     * where ra.assigned_at = (SELECT MAX(assigned_at) FROM referral_assignments WHERE referral_id = ra.referral_id)
+     * and au.id = $USER_ID
+     */
+    fun currentlyAssignedTo(authUserId: String): Specification<Referral> {
```

Comment:
unless you mean the SQL itself?

Reply:
Good idea... I'll run it with debug on to see what the sql output looks like

Reply:
The SQL that is outputted during debug is:

```select
*
from
referral referral0_ inner join referral_assignments assignment1_ on referral0_.id=assignment1_.referral_id
left outer join auth_user authuser2_ on assignment1_.assigned_to_id=authuser2_.id
where assignment1_.assigned_at = (
          select 
               max(assignment4_.assigned_at) 
          from referral referral3_ inner join referral_assignments assignment4_ on referral3_.id=assignment4_.referral_id 
          where referral0_.id=referral3_.id) 
 and authuser2_.id=?

```

It is only slightly different whereby have to do an additional join to `Referral` in the subquery because referral_id is not immediately available on the `ReferralAssignment` entity object but is available in pure SQL.

Reply:
👍